File: Semestrul 4/SDI/lab04/cs/Repository/Impl/ResultRepositoryImpl.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using cs.Domain;
using cs.Repository.Abstracts;
using cs.Repository.Interfaces;

namespace cs.Repository.Impl;

public class ResultRepositoryImpl : AbstractDBRepository<long, Result>, IResultRepository
{
    private readonly IEventRepository _eventRepository;
    private readonly IParticipantRepository _participantRepository;

    public ResultRepositoryImpl(IEventRepository eventRepository, IParticipantRepository participantRepository)
        : base("Results", "id")
    {
        _eventRepository = eventRepository;
        _participantRepository = participantRepository;
    }

    protected override string GetInsertParameters()
    {
        return "event_id, participant_id, discipline_id, points";
    }

    protected override string GetUpdateParameters()
    {
        return "event_id=@eventId, participant_id=@participantId, discipline_id=@disciplineId, points=@points";
    }

    protected override int GetUpdateParameterCount()
    {
        return 4;
    }

    protected override Result CreateEntity(IDataReader reader)
    {
        long id = Convert.ToInt64(reader["id"]);
        long eventId = Convert.ToInt64(reader["event_id"]);
        long participantId = Convert.ToInt64(reader["participant_id"]);
        int disciplineId = Convert.ToInt32(reader["discipline_id"]);
        int points = Convert.ToInt32(reader["points"]);

        Event event_ = _eventRepository.FindOne(eventId);
        Participant participant = _participantRepository.FindOne(participantId);

        Discipline discipline = disciplineId switch
        {
            1 => Discipline.RUNNING,
            2 => Discipline.SWIMMING,
            3 => Discipline.CYCLING,
            _ => throw new DataException($"Unknown discipline ID: {disciplineId}")
        };

        Result result = new Result(event_, participant, discipline, points);
        result.SetId(id);
        return result;
    }

    protected override void SetParameters(SQLiteCommand command, Result entity)
    {
        command.Parameters.AddWithValue("@eventId", entity.GetEvent().GetId());
        command.Parameters.AddWithValue("@participantId", entity.GetParticipant().GetId());

        int disciplineId = entity.GetDiscipline() switch
        {
            Discipline.RUNNING => 1,
            Discipline.SWIMMING => 2,
            Discipline.CYCLING => 3,
            _ => throw new ArgumentException($"Unknown discipline: {entity.GetDiscipline()}")
        };

        command.Parameters.AddWithValue("@disciplineId", disciplineId);
        command.Parameters.AddWithValue("@points", entity.GetPoints());
    }

    public List<Result> FindByEvent(int eventId)
    {
        string sql = "SELECT * FROM Results WHERE event_id = @eventId";
        List<Result> results = new List<Result>();

        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            command.Parameters.AddWithValue("@eventId", eventId);

            using var reader = command.ExecuteReader();
            while (reader.Read())
            {
                results.Add(CreateEntity(reader));
            }
        }
        catch (Exception ex)
        {
            // Logger.LogError($"Error finding results by event ID: {ex.Message}");
        }

        return results;
    }

    public List<Result> FindByParticipant(int participantId)
    {
        string sql = "SELECT * FROM Results WHERE participant_id = @participantId";
        List<Result> results = new List<Result>();

        try
        {
            using var command = new SQLiteCommand(sql, _connection);
            command.Parameters.AddWithValue("@participantId", participantId);

            using var reader = command.ExecuteReader();
            while (reader.Read())
            {
                results.Add(CreateEntity(reader));
            }
        }
        catch (Exception ex)
        {
            // Logger.LogError($"Error finding results by participant ID: {ex.Message}");
        }

        return results;
    }
}

File: Semestrul 4/SDI/lab07/java/TriathlonPersistence/src/main/java/triathlon/persistence/impl/ResultRepositoryImpl.java
package triathlon.persistence.impl;

import triathlon.model.*;
import triathlon.persistence.interfaces.*;
import triathlon.persistence.abstracts.AbstractDBRepository;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class ResultRepositoryImpl extends AbstractDBRepository<Long, Result> implements ResultRepository {
    private final EventRepository eventRepository;
    private final ParticipantRepository participantRepository;

    public ResultRepositoryImpl(EventRepository eventRepository, ParticipantRepository participantRepository) {
        super("Results", "id");
        this.eventRepository = eventRepository;
        this.participantRepository = participantRepository;
    }

    @Override
    protected String getInsertParameters() {
        return "event_id, participant_id, discipline_id, points";
    }

    @Override
    protected String getUpdateParameters() {
        return "event_id=?, participant_id=?, discipline_id=?, points=?";
    }

    @Override
    protected int getUpdateParameterCount() {
        return 4;
    }

    @Override
    protected Result createEntity(ResultSet data) throws SQLException {
        Long id = data.getLong("id");
        Long eventId = data.getLong("event_id");
        Long participantId = data.getLong("participant_id");
        int disciplineId = data.getInt("discipline_id");
        int points = data.getInt("points");

        Event event = eventRepository.findOne(eventId).get();
        Participant participant = participantRepository.findOne(participantId).get();

        Discipline discipline = switch (disciplineId) {
            case 1 -> Discipline.RUNNING;
            case 2 -> Discipline.SWIMMING;
            case 3 -> Discipline.CYCLING;
            default -> throw new SQLException("Unknown discipline ID: " + disciplineId);
        };

        Result result = new Result(event, participant, discipline, points);
        result.setId(id);
        return result;
    }

    @Override
    protected void saveEntity(PreparedStatement preparedStatement, Result entity) throws SQLException {
        preparedStatement.setLong(1, entity.getEvent().getId());
        preparedStatement.setLong(2, entity.getParticipant().getId());

        int disciplineId = switch (entity.getDiscipline()) {
            case RUNNING -> 1;
            case SWIMMING -> 2;
            case CYCLING -> 3;
        };

        preparedStatement.setInt(3, disciplineId);
        preparedStatement.setInt(4, entity.getPoints());
    }

    @Override
    public List<Result> findByEvent(Integer eventId) {
        String sql = "SELECT * FROM Results WHERE event_id = ?";
        List<Result> results = new ArrayList<>();

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setInt(1, eventId);

            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    results.add(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
//            logger.error("Error finding results by event ID: " + e.getMessage());
        }

        return results;
    }

    @Override
    public List<Result> findByParticipant(Integer participantId) {
        String sql = "SELECT * FROM Results WHERE participant_id = ?";
        List<Result> results = new ArrayList<>();

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setInt(1, participantId);

            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    results.add(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
//            logger.error("Error finding results by participant ID: " + e.getMessage());
        }

        return results;
    }
}